src/Control/Comonad/Cofree/Cofreer.hs

```diff
@@ -0,0 +1,88 @@
+{-# LANGUAGE FlexibleInstances, GADTs, MultiParamTypeClasses, TypeFamilies, UndecidableInstances #-}
+module Control.Comonad.Cofree.Cofreer
+( CofreerF(..)
+, Cofreer(..)
+, headF
+, tailF
+, cowrap
+, coiter
+, unfold
+, extract
+, unwrap
+) where
+
+import Control.Arrow ((&&&))
+import Control.Comonad
+import Control.Comonad.Cofree.Class
+import Data.Bifunctor
+import Data.Functor.Classes
+import Data.Functor.Foldable hiding (unfold)
+
+data CofreerF f a b where
+  Cofree :: a -> (x -> b) -> f x -> CofreerF f a b
+
+headF :: CofreerF f a b -> a
+headF (Cofree a _ _) = a
+
+tailF :: Functor f => CofreerF f a b -> f b
+tailF (Cofree _ t r) = t <$> r
+
+
+newtype Cofreer f a = Cofreer { runCofreer :: CofreerF f a (Cofreer f a) }
```

Comment:
This might be a slightly novel presentation. It’s the left Kan extension of Cofree, a la Oleg Kiselyov’s work on Freer.